table.go

```diff
@@ -351,7 +362,140 @@ func (t *table) backfillTo() time.Time {
 }
 
 func (t *table) iterate(ctx context.Context, outFields core.Fields, includeMemStore bool, onValue func(bytemap.ByteMap, []encoding.Sequence) (more bool, err error)) error {
-	return t.rowStore.iterate(ctx, outFields, includeMemStore, onValue)
+	it := &iteration{
+		t:               t,
+		ctx:             ctx,
+		outFields:       outFields,
+		includeMemStore: includeMemStore,
+		onValue:         onValue,
+		errCh:           make(chan error, 1),
+	}
+	t.db.requestedIterations <- it
+	return <-it.errCh
+}
+
+// coalesceIterations coalesces multiple parallel iterations for a single table,
+// which avoids having to do multiple table scans of the same table. It is
+// single-threaded, meaning that each ZenoDB instance performs only one scan at
+// a time. Since ZenoDB instances are meant to be sized small, this shouldn't be
+// a performance issue.
+func (db *DB) coalesceIterations() {
+	for it := range db.requestedIterations {
+		db.coalesceIteration(it)
+	}
+}
+
+func (db *DB) coalesceIteration(it *iteration) {
+	iterations := append([]*iteration(nil), it)
+	iterationsForOtherTables := make([]*iteration, 0)
+
+coalesceLoop:
+	for {
+		select {
+		case it2 := <-db.requestedIterations:
+			if it2.t == it.t {
+				iterations = append(iterations, it2)
+			} else {
+				iterationsForOtherTables = append(iterationsForOtherTables, it2)
+			}
+		case <-time.After(db.opts.IterationCoalesceInterval):
+			// stop waiting to coalesce
+			break coalesceLoop
+		}
+	}
+
+	// re-enqueue iterations for other tables since we won't be handling them
+	// here
+	for _, otherIt := range iterationsForOtherTables {
+		db.requestedIterations <- otherIt
+	}
+
+	var maxDeadline time.Time
+	includeMemStore := false
+	allOutFields := make(core.Fields, 0)
+	hasOutField := func(field core.Field) bool {
+		for _, existingField := range allOutFields {
+			if existingField.String() == field.String() {
+				return true
+			}
+		}
+		return false
+	}
+
+	for _, it := range iterations {
+		includeMemStore = includeMemStore || it.includeMemStore
+		deadline, hasDeadline := it.ctx.Deadline()
+		if hasDeadline && deadline.After(maxDeadline) {
+			maxDeadline = deadline
```

Comment:
It seems a little unusual that the union of deadlines is the latest deadline vs the earliest, but I'm not familiar with why you'd set one or it's implications here.  I guess in this case you don't want something lightweight crowding out something heavy weight with a tighter deadline?

Reply:
Correct. If the caller really cares about its own deadline that much, it'll give up on its own. That should not however cause us to stop iterating for clients that are willing to wait longer.